ambry-tools/src/main/java/com.github.ambry.tools/perf/rest/PerfNioServer.java

```diff
@@ -0,0 +1,419 @@
+      running = false;
+      return shutdownLatch.await(timeout, timeUnit);
+    }
+
+    /**
+     * Initializes load creation by adding slots based on concurrency required.
+     */
+    private void init() {
+      for (int i = 0; i < perfConfig.perfNioServerConcurrency; i++) {
+        slots.add(true);
+      }
+      logger.info("Setup to maintain concurrency of {}", perfConfig.perfNioServerConcurrency);
+    }
+
+    /**
+     * Callback for when response for a request is complete.
+     */
+    private class OperationCallback implements Callback<Void> {
+      private final Logger logger = LoggerFactory.getLogger(getClass());
+
+      /**
+       * Reports exceptions if any and adds a slot so that it can be reused for another request.
+       * @param result The result of the request. This would be non null when the request executed successfully
+       * @param exception The exception that was reported on execution of the request
+       */
+      @Override
+      public void onCompletion(Void result, Exception exception) {
+        slots.add(true);
+        logger.trace("Slot added");
+      }
+    }
+  }
+
+  /**
+   * Perf test implementation of {@link RestRequest}.
+   */
+  private static class PerfRestRequest implements RestRequest {
+    private static final String PERF_PATH = "perf-path";
+    private static final String PERF_URI = "perf-uri";
+    private static final String MULTIPLE_HEADER_VALUE_DELIMITER = ", ";
+
+    private final RestMethod restMethod;
+    private final ReadableStreamChannel readableStreamChannel;
+    private final Map<String, Object> args;
+    private final RestRequestMetricsTracker restRequestMetricsTracker = new RestRequestMetricsTracker();
+
+    /**
+     * Creates an instance of PerfRestRequest.
+     * @param restMethod the {@link RestMethod} desired.
+     * @param usermetadata the usermetadata that needs to be stored with the blob.
+     * @param chunk the data that will form each chunk of this request upto the size defined in {@code totalBlobSize}.
+     * @param totalBlobSize the total size of the blob represented by this PerfRestRequest.
+     */
+    public PerfRestRequest(RestMethod restMethod, String usermetadata, byte[] chunk, long totalBlobSize) {
+      this.restMethod = restMethod;
+      Map<String, Object> inFluxArgs = new HashMap<String, Object>();
+      if (restMethod.equals(RestMethod.POST)) {
+        readableStreamChannel = new PerfRSC(chunk, totalBlobSize);
+        addValueForHeader(inFluxArgs, RestUtils.Headers.BLOB_SIZE, Long.toString(totalBlobSize));
+        addValueForHeader(inFluxArgs, RestUtils.Headers.SERVICE_ID, "PerfNioServer");
+        addValueForHeader(inFluxArgs, RestUtils.Headers.AMBRY_CONTENT_TYPE, "application/octet-stream");
+        addValueForHeader(inFluxArgs, "x-um-perf-user-metadata", usermetadata);
+      } else {
+        readableStreamChannel = new PerfRSC(null, 0);
+      }
+      args = Collections.unmodifiableMap(inFluxArgs);
```

Comment:
A reference to the unmodifiable map is returned on a call to `getArgs()`. Whoever gets the reference should not able to modify it (since it is meant to be read only).